src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/array/element.hpp

```diff
@@ -80,7 +78,7 @@ class BSONCXX_API element : private document::element {
     using document::element::raw;
 
    private:
-    friend class view;
+    friend class ::bsoncxx::v_noabi::array::view;
```

Comment:
> C++ does not provide a mechanism that enforces a friend declaration always refers to an existing declaration, rather than introducing a new declaration

C++11 fixed this by allowing a the friend declaration to name an existing type by omitting the elaborated type tag, e.g.

```suggestion
    friend view;
```

If the name does not name type according to name lookup rules, it generates a hard error (instead of spontaneously generated an incomplete type). Refer to point no. 4 on [cppreference](https://en.cppreference.com/w/cpp/language/friend).

Reply:
TIL! That is very good to know. Converted all friend class declarations into simple type specifier form to enforce existing declaration behavior.

However, for consistency, I've elected to keep the type specifier fully namespace-qualified to avoid possibility for confusion even with this change. e.g.:

```cpp
namespace mongocxx {
inline namespace v_noabi {
namespace options {

class client_session {
  // ...
private:
  friend client_session; // Intent: mongocxx::v_noabi::client_session
                         // Actual: this class (no warnings either...).
};

} // namespace options
} // namespace v_noabi
} // namespace mongocxx
```

There are several other variations of this potential mistake throughout the codebase due to how often we reuse the same identifier within different namespaces.

Full namespace qualification will also help disambiguate entities declared in different ABI namespaces once changes related to CXX-2745 are implemented (`v_noabi`, `v1`, etc.). If including the root namespace is too excessive (`::bsoncxx::`, `::mongocxx::`), then I can at least remove those.